File: src/lib/utils/calendar.ts
import * as types from "../../types";
import * as store from "../store";

export async function getDataForDays(
    size: number,
    year: number,
    month: number,
): Promise<types.db.Entry[]> {
    const data: types.db.Entry[] = [];
    const weekStart = store.obj.get("week-start")!;

    for (let i = 0; i < size; i++) {
        const date = new Date(year, month, i + 1 - getStartDay(year, month, weekStart));

        data.push({
            year: date.getFullYear(),
            month: date.getMonth(),
            date: date.getDate(),
            shift: calcShiftForDay(date),
            note: "",
        });
    }

    return data;
}

function getStartDay(year: number, month: number, weekStart: types.calendar.WeekStart): number {
    // NOTE: if month starts on sunday (0), but the week start is
    //       set to monday (1), than set it to 6 (sunday becomes 6)
    const date = new Date(year, month, 1);
    const weekDay = date.getDay();

    if (weekStart === 0) return weekDay;
    return weekDay === 0 ? 6 : weekDay - 1;
}

export function calcShiftForDay(current: Date): types.calendar.Shift | null {
    const settings = store.obj.get("settings")!;
    if (!settings.startDate || !settings.rhythm.length) return null;

    const sDate = new Date(settings.startDate);
    const diffInDays = Math.round((current.getTime() - sDate.getTime()) / (1000 * 60 * 60 * 24));

    if (diffInDays <= 0) {
        return (
            settings.shifts.find(
                (shift) =>
                    shift.id ===
                    settings.rhythm[
                        settings.rhythm.length +
                            (diffInDays % settings.rhythm.length || -settings.rhythm.length)
                    ],
            ) || null
        );
    }

    return (
        settings.shifts.find(
            (shift) => shift.id === settings.rhythm[diffInDays % settings.rhythm.length],
        ) || null
    );
}

File: internal/calendar/calendar.go
package calendar

import (
	"time"

	"github.com/knackwurstking/shift-scheduler/internal/types"
)

// TODO: Need a test here, with weekStart set to 0 and 1 (sunday and monday)
func GetDataForDays(year int, month time.Month, weekStart time.Weekday) (data [42]types.DBEntry) {
	data = [42]types.DBEntry{}

	t := time.Date(year, month, 1, 0, 0, 0, 0, time.Local)
	startDay := int(getStartDay(t, weekStart))

	for n := range 42 {
		current := t.AddDate(0, 0, n-startDay)

		data[n] = types.DBEntry{
			Year:  current.Year(),
			Month: current.Month(),
			Day:   current.Day(),
			Shift: nil,
			Note:  "",
		}
	}

	return data
}

func getStartDay(t time.Time, weekStart time.Weekday) time.Weekday {
	// NOTE: Ok, this is still weird to me :(
	weekDay := t.Weekday()

	// In this world the week always starts at sunday not monday
	if weekStart == time.Sunday {
		return weekDay
	}

	if weekDay == time.Sunday {
		return time.Saturday // 6
	} else {
		return weekDay - 1
	}
}

//export function calcShiftForDay(current: Date): Shift | null {
//    const startDate = store.obj.get("startDate")!;
//    const shifts = store.obj.get("shifts")!;
//    const rhythm = store.obj.get("rhythm")!;
//
//    if (!startDate || !rhythm.length) return null;
//
//    const sDate = new Date(startDate);
//    const diffInDays = Math.round(
//        (current.getTime() - sDate.getTime()) / (1000 * 60 * 60 * 24),
//    );
//
//    if (diffInDays <= 0) {
//        return (
//            shifts.find(
//                (shift) =>
//                    shift.id ===
//                    rhythm[
//                        rhythm.length +
//                            (diffInDays % rhythm.length || -rhythm.length)
//                    ],
//            ) || null
//        );
//    }
//
//    return (
//        shifts.find(
//            (shift) => shift.id === rhythm[diffInDays % rhythm.length],
//        ) || null
//    );
//}
